core/src/main/scala/org/apache/spark/sql/SnappyParser.scala

```diff
@@ -587,9 +623,20 @@ class SnappyParser(session: SnappySession)
           .getOrElse(base)
       val expressions = p.asInstanceOf[Seq[Expression]]
           .map(UnresolvedAlias(_, None))
-      val withProjection = g.asInstanceOf[Option[Seq[Expression]]]
-          .map(Aggregate(_, expressions, withFilter))
-          .getOrElse(Project(expressions, withFilter))
+      val gr = g.asInstanceOf[Option[(Seq[Expression], Seq[Seq[Expression]], Option[String])]]
+      val withProjection = gr.map(x => {
+        x._3.get match {
+          // group by cols with rollup
+          case "ROLLUP" => Aggregate(Seq(Rollup(x._1)), expressions, withFilter)
+          // group by cols with cube
```

Comment:
Instead of this matching etc, a cleaner approach is to have groupBy and cubeRollUpGroupingSet return a Rule1 of function (Seq[Expression], LogicalPlan) => LogicalPlan

So then groupBy can be like below when cubeRollUpGroupingSet likewise returns Rule1 of (Seq[Expression], Seq[Expression], LogicalPlan) => LogicalPlan:

```scala
protected final def groupBy: Rule1[(Seq[Expression], Seq[Seq[Expression]],  Option[String])] = rule {
  GROUP ~ BY ~ (expression + (',' ~ ws)) ~ cubeRollUpGroupingSet.? ~>{ (exprs: Any, g: Any) =>
    (expressions: Seq[Expression], plan: LogicalPlan) =>
      val groupByExprs = exprs.asInstanceOf[Seq[Expression]]
      g.asInstanceOf[Option[(Seq[Expression], Seq[Expression], LogicalPlan) => LogicalPlan] match {
        case Some(crgs) =>crgs(groupByExprs, expressions, plan)
        case None => Aggregate(groupByExprs, expressions, plan)
      }
  }
}
```